plugins/catalog-react/src/components/EntityOwnerPicker/EntityOwnerPicker.tsx

```diff
@@ -67,6 +76,57 @@ export const EntityOwnerPicker = () => {
     queryParamOwners.length ? queryParamOwners : filters.owners?.values ?? [],
   );
 
+  const {
+    loading,
+    error,
+    value: ownerEntities,
+  } = useAsync(async () => {
+    const availableOwners = [
+      ...new Set(
+        backendEntities
+          .flatMap((e: Entity) =>
+            getEntityRelations(e, RELATION_OWNED_BY).map(o =>
+              stringifyEntityRef(o),
+            ),
+          )
+          .filter(Boolean) as string[],
+      ),
+    ];
+    const { items } = await catalogApi.getEntitiesByRefs({
+      entityRefs: availableOwners,
+      fields: [
+        'kind',
+        'metadata.name',
+        'metadata.title',
+        'metadata.namespace',
+        'spec.profile.displayName',
+      ],
+    });
+    return availableOwners
+      .map(
+        (e, i) =>
+          items[i] || ({ metadata: { name: e }, kind: 'Group' } as Entity),
```

Comment:
`e` isn't a name, it's an entire entity ref isn't it.

This code gets a little confusing since it kinda mixes refs and labels and entire entities to some degree. Would it help to do

- rename the first variable to `ownerEntityRefs`
- rename the second variable to `ownerEntitiesOrNull`
- make an intermediate variable named something like `owners` with the type `Array<{ label: string, entityRef: string }>`, where the label has been set to whatever seemed to be correct for that item
- then finally sort and return that intermediate variable

Now the code outside of the useAsync will have all items nice and sorted, and having easy access to both display labels and entity refs as needed within each entity. Importantly, the code below can also then work _entirely_ with entity refs as the main vehicle of conveying data, and the labels only serve as a side thing for display purposes. `selectedOwners` should _not_ contain `humanizeEntity`'d strings; it should contain entity refs. The autocomplete would likely have for example `getOptionLabel={option => option.label}` since that one was already precomputed inside the useAsync.

Reply:
Definitely agree, updated with the new variable names. Initially, I was trying to update the code to use `Entities` as the primary data array but it does just make everything confusing.

I think this is related to my question above, but we're using the `humanizeEntityRef`'d version of the entity ref in `selectedOwners`. Should that logic also be updated?

Reply:
Looks completely doable to use full entity refs for the internal state with a humanized version shown in the URL. That being said, I have 2 concerns,
1.  `humanizeEntityRef` diverging from being parseable back into a full entity ref.
2. The [EntityOwnerFilter check](https://github.com/sennyeya/backstage/blob/4df9b603ebd9cd7b8bae1fc7588db863bfde5b93/plugins/catalog-react/src/filters.ts#L126) is using a string equality test against the `humanizeEntityRef`'d entity ref, which doesn't seem great.